src/ChangeStream/ChangeStream.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+/*
+ * This file is part of the puli/repository package.
+ *
+ * (c) Bernhard Schussek <bschussek@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Puli\Repository\ChangeStream;
+
+use Puli\Repository\Api\Resource\PuliResource;
+use Puli\Repository\Api\ResourceRepository;
+use Puli\Repository\ChangeStream\Normalizer\DirectoryResourceNormalizer;
+use Puli\Repository\ChangeStream\Normalizer\FileResourceNormalizer;
+use Puli\Repository\ChangeStream\Normalizer\GenericResourceNormalizer;
+use Puli\Repository\ChangeStream\Normalizer\LinkResourceNormalizer;
+use Puli\Repository\ChangeStream\Normalizer\ResourceNormalizer;
+
+/**
+ * Stream to track repositories changes and fetch previous versions of resources.
+ *
+ * @since  1.0
+ *
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class ChangeStream
+{
+    /**
+     * @var array
+     */
+    private $stack;
+
+    /**
+     * @var ResourceNormalizer[]
+     */
+    private $normalizers;
+
+    /**
+     * Create a change stream.
+     */
+    public function __construct()
+    {
+        $this->stack = array();
+
+        $this->normalizers = array(
+            new LinkResourceNormalizer(),
+            new DirectoryResourceNormalizer(),
+            new FileResourceNormalizer(),
+            new GenericResourceNormalizer(),
```

Comment:
That exactly is why PuliResource extends Serializable :) People can adjust the serialization logic in their own implementations. If they need to inject external dependencies, we can add an initializer mechanism (similar to BindingInitializer) later on.